### PR TITLE
Progress fix

### LINKED ISSUE #1937: progress bar does not stop when Ctrl+C is pressed
Emulating Ctrl+C sent to a process is not exactly trivial, but this presumably does the trick:
```
import signal
import subprocess
import time
def test():
    proc = subprocess.Popen(["python", "-c", "import datatable as dt; DT = dt.fread('/very/large/file')"],
                            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    time.sleep(0.1)
    proc.send_signal(signal.Signals.SIGINT)
    ret = proc.wait()
    print("Return code = %r" % ret)
    print("STDOUT = %r" % proc.stdout.read().decode())
    print("STDERR = %r" % proc.stderr.read().decode())

test()
```

### PR DESCRIPTION
**Pull request by oleksiyskononenko**
- fix progress bar not stopping when Ctrl+C is pressed;
- make parallel infrastructure more interrupt friendly;
- improve progress bar reporting;
- add tests, progress bar may need some additional tests though;
- adjust progress reporting for FTRL model.

Closes #1937

### CODE REVIEW

File: c/parallel/parallel_for_ordered.cc [hunk 1]
```diff
@@ -300,23 +300,29 @@ void ordered::parallel(function<void(size_t)> pre_ordered,
                        function<void(size_t)> post_ordered)
 {
   if (sch->n_threads <= 1) {
+    enable_monitor(true);
```
Comment (st-pasha) [inline c/parallel/parallel_for_ordered.cc]: What if there is an exception during the ordered loop execution -- will the monitor remain enabled, potentially later causing a crash in python?

File: c/parallel/api.h [hunk 2]
```diff
@@ -101,7 +112,7 @@ void barrier();
  * Run parallel loop `for i in range(nrows): f(i)`, with dynamic scheduling.
  */
 void parallel_for_dynamic(size_t nrows, std::function<void(size_t)> fn);
-void parallel_for_dynamic(size_t nrows, size_t nthreads,
+void parallel_for_dynamic(size_t nrows, NThreads _NThreads,
```
Comment (st-pasha) [inline c/parallel/api.h]: I don't think that's a valid variable name in C++, see https://stackoverflow.com/questions/228783/what-are-the-rules-about-using-an-underscore-in-a-c-identifier

File: c/parallel/ztest_parallel_for.cc [hunk 3]
```diff
@@ -102,7 +102,7 @@ void test_parallel_for_ordered(size_t n) {
 
   dt::parallel_for_ordered(
     /* n_iters = */ n,
-    /* n_threads = */ size_t(-1),
+    /* n_threads = */ dt::NThreads(size_t(-1)),
```
Comment (st-pasha) [inline c/parallel/ztest_parallel_for.cc]: this is same as `dt::NThreads()`

File: c/progress/progress_bar.cc [hunk 4]
```diff
@@ -32,65 +32,71 @@ static void init_status_strings() {
...
 
+progress_bar::~progress_bar() {}
 
-progress_bar::progress_bar() {
+
+void progress_bar_disabled::set_progress(double, double) noexcept {};
+void progress_bar_disabled::set_status_finished() {};
+void progress_bar_disabled::set_status_error(bool) {};
+void progress_bar_disabled::set_message(std::string&&) {};
+void progress_bar_disabled::refresh() {};
```
Comment (st-pasha) [inline c/progress/progress_bar.cc]: semicolons not needed after function bodies

File: c/parallel/api.h (hunk 2 above)
Comment (oleksiyskononenko) [reply c/parallel/api.h]: Changed it everywhere to `NThreads_` to be safe.

File: c/parallel/ztest_parallel_for.cc (hunk 3 above)
Comment (oleksiyskononenko) [reply c/parallel/ztest_parallel_for.cc]: Yeah, didn't notice this was just a way to get all the available threads. Expected to see something like `dt::num_threads_in_pool()` here. Anyways, simplified this to `dt::NThreads()`.

File: c/progress/progress_bar.cc (hunk 4 above)
Comment (oleksiyskononenko) [reply c/progress/progress_bar.cc]: Not sure how they got there :) Fixed.

File: c/parallel/parallel_for_ordered.cc (hunk 1 above)
Comment (oleksiyskononenko) [reply c/parallel/parallel_for_ordered.cc]: Yes, could be a problem. Also for `parallel_for_static()` when we don't start a parallel region. Can add `try`/`catch` around the `func(i)` call and stop monitor thread if there is an exception, but not sure what to do with the exception itself in this case.

Comment (st-pasha) [reply c/parallel/parallel_for_ordered.cc]: well, you can always just re-throw it

Comment (oleksiyskononenko) [reply c/parallel/parallel_for_ordered.cc]: Sounds like a plan :)

Comment (st-pasha) [reply c/parallel/parallel_for_ordered.cc]: I moved try-catch blocks out of the loop: seem more robust that way.

File: c/progress/progress_manager.h [hunk 5]
```diff
@@ -53,6 +56,20 @@ class progress_manager {
...
+    // the interrupt, it also resets `caught_interrupt` back to `false`.
+    mutable bool caught_interrupt;
+
+    // This flag is kind of similar to the above, but is atomic and not
+    // cleared by the interrupt handler. It is only cleared when threads are
+    // joining, and used in the cases when there is no special scheduler
+    // (like in the cases of `parallel_for_static`, `nested_for_static`
+    // or a single threaded `parallel_for_ordered`) but we still want
+    // to abort execution of all threads when SIGINT is received.
+    mutable std::atomic<bool> abort_execution;
```
Comment (st-pasha) [inline c/progress/progress_manager.h]: I'm still not sure I understand the purpose of this flag fully.
I see that it is set together with `caught_interrupt`, but cleared at a later stage. But then, why do we need to clear `caught_interrupt` so early? Would it work if we removed `caught_interrupt` altogether, and left just this flag?
What about other, non-interrupt exceptions?

File: c/parallel/parallel_for_ordered.cc (hunk 1 above)
Comment (oleksiyskononenko) [reply c/parallel/parallel_for_ordered.cc]: Yes, though I'm not sure if anything besides the `*_ordered()` functions can throw. May still be a little bit better from the performance point of view.

File: c/progress/progress_manager.h (hunk 5 above)
Comment (oleksiyskononenko) [reply c/progress/progress_manager.h]: Replaced these two flags with just one, hopefully its purpose is clear now.

**Pull request merged**